oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsSoundCard.java

```diff
@@ -0,0 +1,140 @@
+                    , WmiUtil.getString(SOUND_CARD_QUERY_RESULT, SoundCardName.MANUFACTURER, i));
+        }
+        createClause(NAME_MAP);
+
+    }
+
+    /**
+     * Creates our Win32_PnPSignedDevice query with the WHERE clause taking the attributes
+     * from our map.
+     * @param map
+     *       The map whose keys will be used inside the Where Clause.
+     */
+    private static void createClause(Map<String, String> map) {
+        boolean isEnd = true;
+        for (String key : map.keySet()) {
+            if (isEnd) {
+                whereClause += "\"%" + key + "%\"";
+                isEnd = false;
+            } else {
+                whereClause += " OR DeviceName LIKE";
+                whereClause += "\"%" + key + "%\"";
+            }
+        }
+    }
+
+
+    /**
+     * Short method to build our kernel string.
+     * @param index
+     *       The row to fetch data.
+     * @param cardKernelQueryResult
+     *       The result to take data from.
+     * @return
+     *       The kernel string.
+     */
+    private static String getAudioCardKernelVersion(int index, WbemcliUtil.WmiResult cardKernelQueryResult) {
+        String audioCardKernel;
+        audioCardKernel =
+                WmiUtil.getString(cardKernelQueryResult,SoundCardKernel.DRIVERPROVIDERNAME, index) + " "
+                        + WmiUtil.getString(cardKernelQueryResult, SoundCardKernel.DEVICENAME, index) + " "
+                        + WmiUtil.getString(cardKernelQueryResult, SoundCardKernel.DRIVERNAME, index) + " "
+                        + WmiUtil.getString(cardKernelQueryResult, SoundCardKernel.DRIVERVERSION, index);
+        return audioCardKernel;
+    }
+
+
+    /**
+     * Does the following :
+     * <ul>
+     *     <li>Creates and runs our 'built where clause' query</li>
+     *     <li>Then iterates over the key values of our map and compares the key
+     *     with the DeviceName attribute returned by our where-clause query.If its a
+     *     match then we create our SoundCard object
+     *     </li>
+     * </ul>
+     * <br>
+     * NOTE : The reason why the codec name is same as the card name is because windows does not provide
+     * the name of the codec chip but sometimes the name of the card returned is infact the name of the codec
+     * chip also.
+     * Example : Realtek ALC887 HD Audio Device
+     * @return List of sound cards
+     */
+    public static List<WindowsSoundCard> getSoundCards() {
+        WbemcliUtil.WmiQuery<SoundCardKernel> cardKernelQuery = new WbemcliUtil.WmiQuery<>(driverQuery +" "+ whereClause, SoundCardKernel.class);
+        WbemcliUtil.WmiResult<SoundCardKernel> cardKernelQueryResult = WmiUtil.queryWMI(cardKernelQuery);
+
+        List<WindowsSoundCard> soundCards = new ArrayList<>();
+
+        for (Map.Entry<String, String> nameSet : NAME_MAP.entrySet()) {
+            for (int i = 0; i < cardKernelQueryResult.getResultCount(); i++) {
```

Comment:
This works but it's O(n^2).  It seems less convoluted to make the `WmiResult` iteration the outer loop and then just get from the map (careful with nulls, see `MapUtil if you need some defaults) so the whole loop is just O(n).

Reply:
`...driverQuery = "Win32PnPSignedDriver WHERE DeviceName LIKE"  `, like this right ? This looks better instead of using two different variables.

Reply:
You can stick with one single variable, yes... just initialize empty and build it all in one place so someone editing this code later doesn't have to hunt down for various initializations in various places.

Reply:
Well I managed to remove the outer loop , kept the inner loop as the main loop and used NAME_MAP.containsKey() inside the loop instead of iterating the keys one by one and checking. Pushed the changes just now..

Reply:
I guess this is fine ?